### PR TITLE
Fix #622 - Add a custom trip planning API URL setting

### LINKED ISSUE #622: Add a custom trip planning API URL setting
**Summary:** 

For other regions that have OneBusAway deployed, it would be nice if there was a way they could test an OTP server with their region, without us updating the Regions API in production.  For example, Puget Sound has an OTP server, but currently there isn't a good way of anyone in Puget Sound testing the system without going straight to their OTP URL in production or forking the app.  This would also allow regions to test new OTP server configurations (e.g., addition of real-time data or upgrades) before going to production.  

This setting is similar to the current "Custom OBA Server API" setting that we have, but for OTP.  I imagine this would also go under the "Advanced" preference category.

### PR DESCRIPTION
**Pull request by cagryInside**
Added a new field to set OTP API Url.

**Screenshots:**
device-2016-07-28-120207 device-2016-07-28-120253

**Behavior:**

1. Redraw navigation drawer after setting the custom url
1. When there is a custom otp url, if a user selects  a new region 
   1. If the region has otp url then custom url will be canceled.
   1. If the region hasn't otp url then custom url will remain.
 
cc'd @barbeau

### CODE REVIEW

File: onebusaway-android/src/main/java/org/onebusaway/android/app/Application.java [hunk 1]
```diff
@@ -281,13 +281,20 @@ public synchronized ObaRegion getCurrentRegion() {
     }
 
     public synchronized void setCurrentRegion(ObaRegion region) {
+        setCurrentRegion(region, true);
+    }
+
+    public synchronized void setCurrentRegion(ObaRegion region, boolean didRegionChanged) {
```
Comment (barbeau) [inline onebusaway-android/src/main/java/org/onebusaway/android/app/Application.java]: Could you change the variable name to just `regionChanged`?

File: onebusaway-android/src/main/java/org/onebusaway/android/directions/util/TripRequestBuilder.java [hunk 2]
```diff
@@ -236,7 +239,23 @@ public TripRequest execute(Activity activity) {
         request.setShowIntermediateStops(true);
 
         // TripRequest will accept a null value and give a user-friendly error
-        String otpBaseUrl = Application.get().getCurrentRegion().getOtpBaseUrl();
+        String otpBaseUrl;
+        Application application = Application.get();
```
Comment (barbeau) [inline onebusaway-android/src/main/java/org/onebusaway/android/directions/util/TripRequestBuilder.java]: Small tweak - could you shorten variable name to just `app`?

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/HomeActivity.java [hunk 3]
```diff
@@ -389,6 +391,15 @@ protected void onSaveInstanceState(Bundle outState) {
         }
     }
 
+//    @Override
```
Comment (barbeau) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/HomeActivity.java]: Looks like this is commented out - should it just be removed?

File: onebusaway-android/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -564,11 +564,16 @@
     <string name="preferences_arrival_info_style_options_a">OneBusAway classic</string>
     <string name="preferences_arrival_info_style_options_b">Cards</string>
 
+    <string name="preferences_otp_api_servername_title">Trip Planner API Server</string>
```
Comment (barbeau) [inline onebusaway-android/src/main/res/values/strings.xml]: Could you please change "Trip Planner" to "OpenTripPlanner" in both title and summary?

**Comment by barbeau** [comment]
Thanks @cagryInside!  Looks pretty good - I haven't had a chance to test yet, but a few comments inline from looking at code.

**Comment by cagryInside** [comment]
> Thanks @cagryInside! Looks pretty good - I haven't had a chance to test yet, but a few comments inline from looking at code.

@barbeau just pushed another commit that fixed inline comments.

**Comment by cagryInside** [comment]
@barbeau Also added the old url version feature to this PR.

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java [hunk 5]
```diff
@@ -259,44 +266,62 @@ private void changePreferenceSummary(String preferenceKey) {
         if (preferenceKey.equalsIgnoreCase(getString(R.string.preference_key_region))
                 || preferenceKey.equalsIgnoreCase(getString(R.string.preference_key_oba_api_url))) {
             if (Application.get().getCurrentRegion() != null) {
-                regionPref.setSummary(Application.get().getCurrentRegion().getName());
-                customApiUrlPref
+                mPreference.setSummary(Application.get().getCurrentRegion().getName());
+                mCustomApiUrlPref
                         .setSummary(getString(R.string.preferences_oba_api_servername_summary));
+                String customOtpApiUrl = Application.get().getCustomOtpApiUrl();
+                if (customOtpApiUrl == null) {
```
Comment (barbeau) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java]: Looking at the code below line 289, I think this should be `if (!TextUtils.isEmpty(customOtpApiUrl))`?

**Comment by barbeau** [comment]
Thanks @cagryInside!  One comment inline.

Also, I can't seem to clear the custom OTP API after it's set - not sure if it's related to the same comment.  So if I do the following:

1. Install app, region is auto-set to Tampa
1. Set custom OTP URL to https://maps.usf.edu/otp
1. Change region to Puget Sound
1. Set custom OTP URL to an empty string (put cursor in window and backspace to remove all characters)

...I still see the "Plan a trip (beta)" option in the navigation drawer, and if I try to plan a trip, I get "Planning not possible - no region selected" error message:

image

I would expect it to clear the custom OTP API, and hide the nav. drawer option to "Plan a trip".

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java (hunk 5 above)
Comment (cagryInside) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java]: @barbeau do you mean line 273 instead of 289

Comment (barbeau) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java]: Yes, I'm asking about line 273, based on how line 289 looks:

Line 289:
~~~
 +            if (!TextUtils.isEmpty(otpApiUrl)) {
 +                mCustomOtpApiUrlPref.setSummary(otpApiUrl);
 +            } else {
 +                mCustomOtpApiUrlPref.setSummary(
 +                        getString(R.string.preferences_otp_api_servername_summary));
 +            }
~~~

Line 273:
~~~
+                if (customOtpApiUrl == null) {
+                    mCustomOtpApiUrlPref
 +                            .setSummary(getString(R.string.preferences_otp_api_servername_summary));
 +                } else {
 +                    mCustomOtpApiUrlPref.setSummary(customOtpApiUrl);
 +                }
~~~

Seems like these are attempting doing the same thing, but one is using a null check, and the other is using `TextUtils.isEmpty()`.  So I believe these should both be the same, unless I'm missing something.

**Comment by cagryInside** [comment]
> Yes, I'm asking about line 273, based on how line 289 looks:

Yes thats correct, I pushed another commit that fixes this issue and the empty string issue.

**Comment by barbeau** [comment]
Thanks @cagryInside, this looks good!  Great to both support older OTP server deployments, and to add the custom API setting for OTP servers. :+1: :+1:

**Pull request merged**